### PR TITLE
feat(server): Add support for client-side hashing

### PR DESCRIPTION
**Pull request by etnoy**
This is the first major PR related to client side hashing. Here, we focus on getting the server-side plumbing done. 

* New server endpoint for checking for duplicates via checksum
* No changes to how mobile uploads work, future PRs will be able to improve dupe handling with checksums
* Removed the duplication checking when uploading via web, dedupes will be handled server side only in this case
* Updating checksum column to be non-nullable, this was determined to be safe: https://discordapp.com/channels/979116623879368755/1063891307149414410/1088819797518712882
* CLI uploads will require the hash-enabled CLI version: https://github.com/immich-app/CLI/tree/hash

### CODE REVIEW

**Comment by bo0tzz** [comment]
> Mobile apps don't do hashing yet

Is it worth doing that in a separate PR, to keep review sizes down?

File: server/libs/infra/src/db/entities/asset.entity.ts
Comment (michelheusschen) [inline server/libs/infra/src/db/entities/asset.entity.ts:73]: Removing null doesn't make the column nullable, you have change that in the `@Column` decorator and for this specific instance the `@Index` also. But is this change backwards compatible?

File: server/libs/domain/src/media/media.service.ts
Comment (bo0tzz) [reply server/libs/domain/src/media/media.service.ts:1]: I don't think we need to change the paths in this PR at all, and instead leave that to #2047

File: server/libs/infra/src/db/migrations/1679572446996-DeviceIdOptional.ts
Comment (etnoy) [reply server/libs/infra/src/db/migrations/1679572446996-DeviceIdOptional.ts:1]: I could need some help in getting the migrations right. The goal is to make deviceId and deviceAssetId optional (ie remove the NOT NULL requirement)

File: server/libs/infra/src/db/entities/asset.entity.ts
Comment (etnoy) [reply server/libs/infra/src/db/entities/asset.entity.ts:73]: Thanks, can you look at the new version please?

File: server/libs/domain/src/media/media.service.ts
Comment (etnoy) [reply server/libs/domain/src/media/media.service.ts:1]: Never mind, I was confused about the diff view.

These changes are needed because deviceId is no longer available here. I belive the impact will be minimal, only new items will get the new path.

File: server/libs/infra/src/db/entities/asset.entity.ts
Comment (michelheusschen) [reply server/libs/infra/src/db/entities/asset.entity.ts:73]: Yes that's good, you can also remove the `@Index` because checksum can never be null

File: server/libs/infra/src/db/migrations/1679572446996-DeviceIdOptional.ts
Comment (etnoy) [reply server/libs/infra/src/db/migrations/1679572446996-DeviceIdOptional.ts:1]: Done

File: server/libs/infra/src/db/entities/asset.entity.ts
Comment (michelheusschen) [reply server/libs/infra/src/db/entities/asset.entity.ts]: Sorry, I meant to say remove the where part, so you just have `@Index()`. There is still a composite index generated for ownerId and checksum, but it's probably safer to add the index back.

File: server/libs/infra/src/db/migrations/1679572446996-DeviceIdOptional.ts
Comment (michelheusschen) [reply server/libs/infra/src/db/migrations/1679572446996-DeviceIdOptional.ts:1]: Make sure to reflect making these columns nullable in DTOs and other places where it's used

Also, always check the generated migrations, sometimes it creates weird migrations. Now it drops `UQ_userid_checksum` only to recreate it right after, not a big issue now since it's just a constraint but sometimes it does that to columns.

File: server/libs/domain/src/media/media.service.ts
Comment (bo0tzz) [reply server/libs/domain/src/media/media.service.ts:1]: The PR I mentioned also removes deviceId. At the moment this diff is necessary for this branch to build of course, but once that PR merges it should replace this change IMO.

Comment (etnoy) [reply server/libs/domain/src/media/media.service.ts:1]: I discussed with @jrasm91 and this PR will be a band-aid fix to get things going, the real change will be in #2047. 

We could also skip changing media service here, but then thumbs will be in `thumbs/[Object object]` and originals will be in `originals/undefined` :)

File: server/libs/infra/src/db/entities/asset.entity.ts
Comment (etnoy) [reply server/libs/infra/src/db/entities/asset.entity.ts]: After a discussion in discord I think we keep the composite index only

File: server/libs/infra/src/db/migrations/1679662469076-RemoveChecksumIndex.ts
Comment (michelheusschen) [inline server/libs/infra/src/db/migrations/1679662469076-RemoveChecksumIndex.ts:8]: I'm guessing you manually removed the `ADD CONSTRAINT "UQ_userid_checksum"`? You should also remove this query, otherwise you're only removing the constraint here and not adding it back later.

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset-repository.ts:288]: Probably better to filter out null values if there's no reason to keep them included

File: server/apps/immich/src/api-v1/asset/dto/check-existence-of-assets.dto.ts
Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/dto/check-existence-of-assets.dto.ts:10]: Thanks

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:470]: I'm new to ts, but wouldn't that block for every array element? I thought this is faster since it just blocks once, allowing all elements to finish asynchronously

Comment (michelheusschen) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:470]: No it wouldn't block for every element. The repository will make one database call and return all elements. You can see the return type for getAssetsByChecksums is `Promise<AssetEntity[]>`. You can just await the Promise and the type will become `AssetEntity[]`.

If the return type would be `Promise<AssetEntity>[]` then you have to await each individual element.

File: server/libs/infra/src/db/migrations/1679662469076-RemoveChecksumIndex.ts
Comment (etnoy) [reply server/libs/infra/src/db/migrations/1679662469076-RemoveChecksumIndex.ts:8]: We probably need to regenerate the migrations for this PR as they became such a mess. I might need some help on that when we approach merge tiem

Comment (michelheusschen) [reply server/libs/infra/src/db/migrations/1679662469076-RemoveChecksumIndex.ts]: I've updated the migrations

File: web/src/lib/utils/file-uploader.ts
Comment (etnoy) [reply web/src/lib/utils/file-uploader.ts:1]: File-uploader has now been updated to use this api

File: server/libs/infra/src/migrations/1679662469076-RemoveChecksumIndex.ts
Comment (michelheusschen) [inline server/libs/infra/src/migrations/1679662469076-RemoveChecksumIndex.ts:1]: Please merge all migration files into one

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset.service.ts:139]:
```suggestion
        const checksums = [file.checksum, livePhotoFile?.checksum].filter((checksum): checksum is Buffer => !!checksum);
```

File: server/apps/immich/src/api-v1/asset/asset.controller.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset.controller.ts:354]: Why response code `207`?

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset-repository.ts:305]:
```suggestion
    return items.map((asset) => asset.deviceAssetId);
```

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset-repository.ts:301]: The `deviceAssetId` check seems unnecessary, because it's not nullable

**Review by michelheusschen — DISMISSED** [review]
The server implementation seems fine, but for web there are quite some things that would break upload in one way or another.

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/asset-repository.ts:320]: Thanks, removing it

File: web/src/lib/utils/file-uploader.ts
Comment (etnoy) [reply web/src/lib/utils/file-uploader.ts:128]: I discussed the web uploader with Alex and we decided to skip hashing altogether in the web uploader

File: server/libs/infra/src/entities/asset.entity.ts
Comment (jrasm91) [inline server/libs/infra/src/entities/asset.entity.ts:30]: I don't think this change is necessary, it should be the default type.

File: server/libs/infra/src/entities/asset.entity.ts
Comment (jrasm91) [inline server/libs/infra/src/entities/asset.entity.ts:39]: Same here

File: server/libs/infra/src/migrations/1679662469076-RemoveChecksumIndex.ts
Comment (jrasm91) [inline server/libs/infra/src/migrations/1679662469076-RemoveChecksumIndex.ts:1]: There two migration files should be merged, and `remove` should be capitalized (`RemoveChecksum...`)

File: web/src/lib/utils/file-uploader.ts
Comment (jrasm91) [inline web/src/lib/utils/file-uploader.ts:121]: The web no longer does deduping, which I think is fine, since anything besides a hash is less than ideal and hashing in the UI is impractical. However, you've also removed some `addAssetsToAlbum` code. Did you uploading assets directly to an album? I'm assuming this needs to be added back.

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/asset/asset-repository.ts:178]: I don't think this change is necessary, since we still have device asset id everywhere.

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (jrasm91) [inline server/apps/immich/src/api-v1/asset/asset-repository.ts:192]: Same.

**Review by jrasm91 — COMMENTED** [review]
I think the scope of this PR has changed a lot over time. It'd be good to update the description to match what's actually being changed here. From what I can tell it is:
- New server endpoint for checking for duplicates via checksum
- No changes to how mobile uploads work
- Removing unique check from web
- Updating checksum column to be non-nullable

If that is all that is changing, we should remove any unrelated changes, like changes to get the objects and locations for user.

File: server/apps/immich/src/api-v1/asset/asset.controller.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/asset/asset.controller.ts:354]: I think the original idea was to communicate that the response would containe multiple response codes, one for each item. However, we aren't returning multiple _http response_ codes, just our own custom enums, so I think a 200 would probably 
be more appropriate.

Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/asset.controller.ts:354]: Fixed

File: server/apps/immich/src/api-v1/asset/asset-repository.ts
Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/asset-repository.ts:178]: Device asset id isn't null, but it will often be empty. Does this change the equation?

File: web/src/lib/utils/file-uploader.ts
Comment (michelheusschen) [reply web/src/lib/utils/file-uploader.ts:121]: Just below this, the file gets uploaded and `addAssetsToAlbum` is called.  So I assume uploading to albums will still work

Comment (etnoy) [reply web/src/lib/utils/file-uploader.ts:121]: I think Michel is right

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset.service.ts:138]: When uploading a live photo, the duplicate could be either `file` or `livePhotoFile` because the return order is undefined, no?

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (michelheusschen) [inline server/apps/immich/src/api-v1/asset/asset.service.ts:499]: Is mime check planned for this PR?

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (jrasm91) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:138]: There is an existing bug here, in the scenario where the live photo is a duplicate and the still is not uploaded yet. This code used to look up the still asset by checksum and return it, but since it didn't exist it would throw an EntityNotFound and that asset would fail to backup leading to some sync problem on the mobile app. It's unclear to me what should happen in that situation. I think returning the id to either part is fine. Really, I think, they should be two separate upload requests, which would also solve the issue. That requires mobile app changes and does not seem to be a priority for the mobile devs.

File: web/src/lib/utils/file-uploader.ts
Comment (jrasm91) [reply web/src/lib/utils/file-uploader.ts:121]: Ah ok. This was just another code branch where it would skip the upload and still add to album. No applicable anymore :+1:.

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:499]: The TODO note is there because that's the perfect place to put a pre-upload mime check in the future

File: server/apps/immich/src/api-v1/asset/asset.service.ts
Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:138]: Thanks, Jason. I think we can resolve this in this PR and track the issue somewhere else

Comment (jrasm91) [reply server/apps/immich/src/api-v1/asset/asset.service.ts:138]: Sounds good. I think this does semi-resolve an existing bug too btw.

**Review by jrasm91 — APPROVED** [review]

**Review by michelheusschen — APPROVED** [review]

File: server/apps/immich/src/api-v1/asset/dto/asset-check.dto.ts
Comment (alextran1502) [inline server/apps/immich/src/api-v1/asset/dto/asset-check.dto.ts:7]: What is the would we fill in for the ID for the incoming asset?

Comment (etnoy) [reply server/apps/immich/src/api-v1/asset/dto/asset-check.dto.ts:7]: The client is free to choose any id that makes sense for it

Comment (jrasm91) [reply server/apps/immich/src/api-v1/asset/dto/asset-check.dto.ts:7]: Instead of assuming the return/response order is going to be the same, the client sends an ID to keep track of individual items. The ID is returned back along with the result (accepted/rejected).

Comment (alextran1502) [reply server/apps/immich/src/api-v1/asset/dto/asset-check.dto.ts:7]: Ah got it. Thank you for clarifying!

**Pull request merged**
